src/github.com/getlantern/flashlight/util/http.go

```diff
@@ -2,21 +2,162 @@ package util
 
 import (
 	"crypto/tls"
+	"errors"
 	"fmt"
 	"net"
 	"net/http"
 	"net/url"
 	"time"
 
+	"github.com/getlantern/fronted"
 	"github.com/getlantern/golog"
 	"github.com/getlantern/keyman"
 	"github.com/getlantern/waitforserver"
 )
 
+const (
+	defaultAddr = "127.0.0.1:8787"
+)
+
 var (
 	log = golog.LoggerFor("flashlight.util")
+
+	// This is for doing direct domain fronting if necessary. We store this as
+	// an instance variable because it caches TLS session configs.
+	direct  = fronted.NewDirect()
+	timeout = time.After(60 * time.Second)
 )
 
+// HTTPFetcher is a simple interface for types that are able to fetch data over HTTP.
+type HTTPFetcher interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// NewChainedAndFronted creates a new struct for accessing resources using chained
+// and direct fronted servers in parallel.
+func NewChainedAndFronted() *chainedAndFronted {
+	cf := &chainedAndFronted{}
+	dual := &dualFetcher{cf}
+
+	cf.fetcher = dual
+	cf.dual = dual
+	cf.chained = &chainedFetcher{}
+	return cf
+}
+
+// ChainedAndFronted fetches HTTP data in parallel using both chained and fronted
+// servers.
+type chainedAndFronted struct {
+	fetcher HTTPFetcher
+	chained HTTPFetcher
+	dual    HTTPFetcher
+}
+
+// Do will attempt to execute the specified HTTP request using only a chained fetcher
```

Comment:
Will do thanks @fffw! I also changed direct.go to just use the first
connection that succeeds to keep things simple (there was a bug in the
previous where it wouldn't keep filling the connection queue).
On Sep 29, 2015 6:50 PM, "fffw" <notifications@github.com> wrote:

> In src/github.com/getlantern/flashlight/util/http.go
> <https://github.com/getlantern/lantern/pull/3202#discussion_r40751795>:
>
> > +
> > +	cf.fetcher = dual
> > +	cf.dual = dual
> > +	cf.chained = &chainedFetcher{}
> > +	return cf
> > +}
> > +
> > +// ChainedAndFronted fetches HTTP data in parallel using both chained and fronted
> > +// servers.
> > +type chainedAndFronted struct {
> > +	fetcher HTTPFetcher
> > +	chained HTTPFetcher
> > +	dual    HTTPFetcher
> > +}
> > +
> > +// Do will attempt to execute the specified HTTP request using only a chained fetcher
>
> Revise the comment slightly?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/pull/3202/files#r40751795>.
>